internal/wasm/function_definition.go

```diff
@@ -10,33 +10,40 @@ import (
 //
 // Note: Unlike ExportedFunctions, there is no unique constraint on imports.
 func (m *Module) ImportedFunctions() (ret []api.FunctionDefinition) {
-	for i := range m.FunctionDefinitionSection {
-		d := &m.FunctionDefinitionSection[i]
-		if d.importDesc != nil {
-			ret = append(ret, d)
-		}
+	for i := uint32(0); i < m.ImportFunctionCount; i++ {
+		ret = append(ret, m.FunctionDefinition(i))
 	}
 	return
 }
 
 // ExportedFunctions returns the definitions of each exported function.
 func (m *Module) ExportedFunctions() map[string]api.FunctionDefinition {
 	ret := map[string]api.FunctionDefinition{}
-	for i := range m.FunctionDefinitionSection {
-		d := &m.FunctionDefinitionSection[i]
-		for _, e := range d.exportNames {
-			ret[e] = d
+	for i := range m.ExportSection {
+		exp := &m.ExportSection[i]
+		if exp.Type == ExternTypeFunc {
+			d := m.FunctionDefinition(exp.Index)
+			ret[exp.Name] = d
 		}
 	}
 	return ret
 }
 
-// BuildFunctionDefinitions generates function metadata that can be parsed from
+// FunctionDefinition returns the FunctionDefinition for the given `index`.
+func (m *Module) FunctionDefinition(index Index) *FunctionDefinition {
+	// TODO: lazy initialization per function.
```

Comment:
also left TODO in follow up

Reply:
I started to take a look at this to begin with and I am now wondering if it's worth doing. `buildFunctionDefinitions()` is doing a one-pass linear scan; if we break this down per-index it looks like we'd have to de-compose it into a few methods, but then we'd still have to jump around (and do some mini-linear searches)

correct me if I am wrong 😬  I am thinking about things like:

```
	n, nLen := 0, len(functionNames)
	for i := range m.FunctionDefinitionSection {
		d := &m.FunctionDefinitionSection[i]
		// The function name section begins with imports, but can be sparse.
		// This keeps track of how far in the name section we've searched.
		funcIdx := d.index
		var funcName string
		for ; n < nLen; n++ {
			next := &functionNames[n]
			if next.Index > funcIdx {
				break // we have function names, but starting at a later index.
```

Reply:
another strategy could be to keep a stateful `FunctionDefinitionCollection`; when a given `index > lastIndex` is requested, we advance the internal cursor until `index` (basically it acts more or less like a Python generator) 🤔 this way it is still lazy; worst case, when `index==len-1` it reverts to the full linear scan. 

(a simplistic impl would abuse goroutines and channels but that's rather overkill)

Reply:
before solving this TODO, we should address concurrency concern as in https://github.com/tetratelabs/wazero/pull/1425#issuecomment-1530870598